### PR TITLE
fix: do not encode the known length for bls12_381 public types

### PR DESCRIPTION
**Pull request by joschisan**
This pr unifies both implementations of serde for the bls12_381 public types Scalar, G1Affine, G2Affine into one. 

Furthermore we fix the non-human readable encoding to not encode the length of the array since this is known at deserialisation. This is not a breaking change since we only use this encoding for the DKG and tbs public types. In the DKG the encodings are only used for messages between peers and the tpe public types are only ever encoded in human readable format as we have tested here: https://github.com/fedimint/fedimint/pull/4754

The new serde implementation is identical to the one we are trying to upstream here: https://github.com/zkcrypto/bls12_381/pull/125

### CODE REVIEW

**Review by dpc — DISMISSED** [review]
But this should get @elsirion approval before landing.

File: fedimint-core/src/bls12_381_serde.rs [hunk 1]
```diff
@@ -1,81 +1,74 @@
...
-        byte_array.copy_from_slice(&bytes);
+        serdect::array::deserialize_hex_or_bin(&mut byte_array, d)?;
 
         Option::from(Scalar::from_bytes(&byte_array))
-            .ok_or_else(|| Error::custom("Could not decode compressed G1Affine"))
+            .ok_or(Error::custom("Could not decode scalar"))
     }
 }
 
-macro_rules! impl_serde_g {
```
Comment (elsirion) [inline fedimint-core/src/bls12_381_serde.rs:38]: It's so compact now that the macro was more overhead than worth I guess?

File: fedimint-core/src/bls12_381_serde.rs [hunk 2]
```diff
@@ -1,81 +1,74 @@
...
+        D: Deserializer<'d>,
+    {
+        let mut byte_array = [0; 32];
 
-        byte_array.copy_from_slice(&bytes);
+        serdect::array::deserialize_hex_or_bin(&mut byte_array, d)?;
 
         Option::from(Scalar::from_bytes(&byte_array))
-            .ok_or_else(|| Error::custom("Could not decode compressed G1Affine"))
+            .ok_or(Error::custom("Could not decode scalar"))
```
Comment (elsirion) [inline fedimint-core/src/bls12_381_serde.rs:22]: `Error::custom` is a heavy-ish operation (it allocates), should be only run on-demand by keeping `ok_or_else`.

Screenshot 2024-03-29 at 21-18-49 value rs - source

**Review by elsirion — APPROVED** [review]
Looks good, one nit.

**Review by dpc — APPROVED** [review]

**Review by elsirion — APPROVED** [review]

**Pull request merged**